Show ICD warnings in profile

atdotde opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
Show ICD warnings in info box and in the gas change icons.

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. --> @willemferguson

sfuchs79 commented on profile-widget/diveeventitem.cpp:
```diff
@@ -158,6 +173,10 @@ void DiveEventItem::setupToolTipString()
 			name += tr(" (cyl. %1)").arg(internalEvent->gas.index + 1);
+		icd_data icd_data;
+		if (isobaric_counterdiffusion(lastgasmix, mix, &icd_data))
+			name += tr("\nICD ΔN2/ΔHe=%1/%2=%3%").arg(icd_data.dN2 / 10).arg(-icd_data.dHe / 10).arg((-100 * icd_data.dN2 / icd_data.dHe));
```
Not "N2"! :-)
And I also have doubts about the "delta".
@willemferguson did recently use:
&#916;N&#8322;
`&#916;N&#8322;`
But I don't know if it has to be the same here. But it has to be a "&#8322;"

sfuchs79 commented on CHANGELOG.md:
```diff
@@ -1,3 +1,4 @@
+- Warn in case of isobaric counter diffusion
```
Honestly I think that @willemferguson and your work does deserve a little bit more detailed description in the future release notes!
Maybe also add a comment about the rule of 1/5?

willemferguson replied:
I will insert suitable text in the user manual.

sfuchs79 commented:
Another proposal:
What you now print in the info box is for sure correct but it's somehow following a different layout/strategy compared to what @willemferguson prints into the notes.
I would instead of e.g.:
ICD ΔN₂/ΔHe=13/45=28%
print the following:
ICD: ΔHe=-45% ΔN₂=+13%>+9%
This new proposal is following exactly the way we also have the data in the ICD table.

willemferguson commented:
I really like this. I do think that one should show this only for gas switches:
1) During ascent.
2) Where the gas-from contains helium. The gas-to can be any gas: air, nitrox, not necessary trimix.
3) Where the switch results in an increase in the partial pressure of nitrogen.
Otherwise a lot of extra information is generated that is not really of a lot of use.
Thank you, Robert, I do think this is really useful.

willemferguson commented:
Very often there are changes in the order of 3%, e.g. from fN2=0.60 to fN2=0.63. If there were a limit, I would probably not like to make this limit much larger than 2%. Just my subjective feeling. But I do agree with you that in your example it might be rather pointless. But, in such cases, the change in fN2 or pN2 would really determine the significance, not only the drop in fHe or pHe. But I would agree that a change in fHe <= 1% is unlikely to be meaningful. This is the reason why a human interpretation is important, not just a pure number. Rather give the user a full set of data with which to evaluate a scenario. We are not dealing with an exact science, as in many other dive issues. Rather err on the side of too much information.

willemferguson commented:
If there is an ascent and also a change to a more lean gas (maybe 10/50 to 20/37) this is unambiguous information that would indicate a planned ascent, even if its part of a multilevel dive. The criterion that dN2 should be positive, I think, will ensure that the right cases are selected for showing icd data.

sfuchs79 commented on profile-widget/diveeventitem.cpp:
```diff
@@ -158,6 +173,10 @@ void DiveEventItem::setupToolTipString()
 			name += tr(" (cyl. %1)").arg(internalEvent->gas.index + 1);
+		icd_data icd_data;
+		if (isobaric_counterdiffusion(lastgasmix, mix, &icd_data))
+			name += tr("\nICD ΔHe=%1% ΔN₂=%2%>+%3").arg(icd_data.dHe / 10).arg(icd_data.dN2 / 10).arg(icd_data.dHe / 50);
```
This is imho not 100% correct yet. I see the following result for a gas change from (18/45) to EAN50:
![grafik](https://user-images.githubusercontent.com/25316050/35188412-5fa95a0c-fe35-11e7-808e-23cb97998a6c.png)

- Values all should have nicer format with leading plus sign and with a precision of 0.1 -> "+45.0%"
This unfortunately doesn't work with .arg and I didn't find out how this works in Qt. Any idea?
- Last value has wrong sign and is missing the "%"
- Yes, this is nitpicking but can we put "ICD:"?

Concerning translation: Maybe better only translate the "ICD"?

willemferguson replied:
Great work, @atdotde, I will run this through a bunch of my trimix dives and see how it functions.

willemferguson replied:
From what I can see there are only two issues. And I have a comment. 

1. Comment: The convention for the other items in the info box is to use a colon, e.g. 'PO2: 1.2bar'. Seeing we are using a proportional font, maybe it will be more space-efficient if one uses colons in this line as well, instead of '='. The line is getting quite long.
2. Issue 1: As @sfuchs79 indicated, the sign of the last figure needs attention.
3. Issue 2: See attached dive log. The change from 27/23 to EAN50 does not get an ICD info line. 
(I hope the markdown for a list works for the above paragraph)

This is, if I understand correctly, because isobaric_counterdiffusion() was not triggered. After all the dN2 was zero. I would like the icd information even if the change was within the rule_of_fifths.

The second reason may be because an ascent was not triggered at this gas change. Frequently it is necessary to stop at a gas change, to perform a coordinated gas change with buddy or to pick up a deco cylinder in a cave. So there is a short stop during the gas change. We need to ensure that, in such cases, the ascent still applies. 

[example85.xml.tar.gz](https://github.com/Subsurface-divelog/subsurface/files/1649554/example85.xml.tar.gz)

Thank you again fir this work.

atdotde commented:
The latest update addresses most of your comments: Signs, percent signs, and colons are taken care of although for lack of space I decided against showing + signs. The info box line is now shown whenever the He content decreases. I still do not make any attempt to detect "ascent" (as we saw in the vpm-b discussion this notion is as ambiguous as you want as it depends on the time scale over which you average depths). Also, I decided against showing fractions of percent as we never do this for gas fractions (as the measurement of those probably has errors in this range and it is physiologically irrelevant).

willemferguson commented:
I checked this out quite extensively among my trimix dives and it looks pretty ok. I could not find any erroneous or unusual behaviour. Thanks very much for your time. I would say this pull ready-to-merge.

sfuchs79 commented on profile-widget/diveeventitem.cpp:
```diff
@@ -143,21 +158,30 @@ void DiveEventItem::setupPixmap()
+		bool icd = isobaric_counterdiffusion(lastgasmix, mix, &icd_data);
+		if (icd_data.dHe < 0) {
+			if (icd)
+				name += tr("\nICD:ΔHe=%1% ΔN₂=%2%>%3%").arg(icd_data.dHe / 10).arg(icd_data.dN2 / 10).arg(-icd_data.dHe / 50);
```
Thanks for the ":" but may I ask for a space after the colon as well?
That's not only my personal preference but simply comes from looking at the other info box items we have today.
It's always:
**Whatever: Data**
(with small number of exceptions for completely different items)

![grafik](https://user-images.githubusercontent.com/25316050/35193631-ae389f00-fea5-11e7-8a63-cf57e43380b9.png)

sfuchs79 replied:
Wow, that was quick... thanks!
I'll now also do a quick test but I expect that this is all ok now.

sfuchs79 commented:
One comment about showing fractions of percent mainly also to @willemferguson :
In the ICD table we now have (maybe accidentally) a precision of 0.01 for the gas percentage. Following Roberts arguments this is really much too precise. I would show 0.1 there.
I honestly would show 0.1 for ICD gas percent in both occurrences but I would be also ok with 0.1 in the one place and no fractions at the other. But 0.01 vs. nothing looks strange.

atdotde commented:
And here is the appropriate space.

sfuchs79 commented:
Ups, no, there is still one small issue:
I still see a "warning" icon for the gas change from (50/15) to (18/45) at the beginning of dive at 15m.
Info box and table are ok (nothing printed) but only icon is incorrect.

atdotde commented:
OK, the latest incarnation also checks for He decreasing.

sfuchs79 commented:
My latest functional tests show good results. 👍

willemferguson commented:
I go along with decreasing resolution to 0.1 in the planner.

sfuchs79 commented:
@willemferguson 
You're going to change it to 0.1?
@atdotde 
I found a tricky way to keept the ICD string in the info box as short as possible but to add a precision of 0.1 only if needed plus the [+/-] sign. Looks like:
`struct membuffer mb = {};`
`[...]`
`put_format(&mb, tr("\nICD: ΔHe=%+.3g%% ΔN₂=%+.3g%%>%+.3g%%").toUtf8().constData(),  cd_data.dHe / 10.0, icd_data.dN2 / 10.0, -icd_data.dHe / 50.0);`
`name += QString::fromUtf8(mb.buffer, mb.len);`

neolit123 commented on profile-widget/diveeventitem.cpp:
```diff
@@ -143,21 +158,30 @@ void DiveEventItem::setupPixmap()
+				name += tr("\nICD: ΔHe=%1% ΔN₂=%2%>%3%").arg(icd_data.dHe / 10).arg(icd_data.dN2 / 10).arg(-icd_data.dHe / 50);
+			else
+				name += tr("\nICD: ΔHe=%1% ΔN₂=%2%<%3%").arg(icd_data.dHe / 10).arg(icd_data.dN2 / 10).arg(-icd_data.dHe / 50);
+		}
```
are the above arguments percentages - e.g. `%1%` says `25%` or should it just say `25`?

due to the duplication i would re-write the above if statement as:
```
if (icd_data.dHe < 0) {
	name += tr("\nICD: ΔHe=%1% ΔN₂=%2%%3%4%").arg(icd_data.dHe / 10).arg(icd_data.dN2 / 10)
		.arg(icd ? ">" : "<").arg(-icd_data.dHe / 50);
}
```
this reduces the translations strings and the LOC.

`>` and `<` are mathematical symbols and do not need translation. same goes for `=`.

sfuchs79 replied:
Yes, in fact I really would like to try to further improve the printout. At this moment I can also cover @neolit123 proposal.
But what really give to translation? The string in Lubomirs code example `"\nICD: ΔHe=%1% ΔN₂=%2%%3%4%"` is still very "ugly". Shall we maybe translate only "IDC", "He", "N₂" or "ΔHe" and "ΔN₂" or `"&#916;He"` and `"&#916;N&#8322;"` (used by Willem)?

neolit123 replied:
only `ICD` should be translated; the rest are either chemical symbols or math symbols. but the string loses context if you put only `ICD` for translation, thus my vote is to translate the whole string.

willemferguson commented:
I will change precision to 0.1

atdotde commented:
I will merge this and then @sfuchs79 can add his improved format string as a separate commit. I also cancelled the Mac build since this queue seems to be totally overloaded (and it compiles fine on my mac ;-) )

neolit123 commented on profile-widget/diveeventitem.cpp:
```diff
@@ -86,14 +86,29 @@ void DiveEventItem::setupPixmap()
+			if (icd)
+				setPixmap(EVENT_PIXMAP_BIGGER(":gaschange-trimix-ICD-icon"));
+			else
+				setPixmap(EVENT_PIXMAP_BIGGER(":gaschange-trimix-icon"));
```
i can suggest to use `a ? b : c` for these:
```
setPixmap(EVENT_PIXMAP_BIGGER(icd ? ":gaschange-trimix-ICD-icon" : ":gaschange-trimix-icon"));
```
or:
```
setPixmap(EVENT_PIXMAP_BIGGER(icd ? ":gaschange-trimix-ICD-icon" :
	":gaschange-trimix-icon"));
```
